pkg/providers/vsphere/reconciler/reconciler.go

```diff
@@ -167,6 +173,50 @@ func (r *Reconciler) ValidateMachineConfigs(ctx context.Context, log logr.Logger
 	return controller.Result{}, nil
 }
 
+// ReconcileFailureDomains applies the Vsphere FailureDomain objects to the cluster.
+// It also takes care of  deleting the old Vsphere FailureDomains that are not in in the cluster spec anymore.
+func (r *Reconciler) ReconcileFailureDomains(ctx context.Context, log logr.Logger, spec *c.Spec) (controller.Result, error) {
+	if features.IsActive(features.VsphereFailureDomainEnabled()) && len(spec.VSphereDatacenter.Spec.FailureDomains) > 0 {
+		log = log.WithValues("phase", "reconcileFailureDomains")
+		log.Info("Applying Vsphere FailureDomain objects")
+
+		fd, err := vsphere.FailureDomainsSpec(log, spec)
+		if err != nil {
+			return controller.Result{}, err
+		}
+		if err := serverside.ReconcileObjects(ctx, r.client, fd.Objects()); err != nil {
+			return controller.Result{}, errors.Wrap(err, "applying Vsphere Failure Domain objects")
+		}
+
+		desiredFailureDeploymentNames := collection.MapSet(fd.Groups, func(g vsphere.FailureDomainGroup) string {
+			return g.VsphereDeploymentZone.Name
+		})
+	
+		existingVSphereDeploymentZones := &vspherev1.VSphereDeploymentZoneList{}
+		if err := r.client.List(ctx, existingVSphereDeploymentZones,
+			ctrlClient.MatchingLabels{vsphere.VsphereDataCenterConfigNameLabel: spec.VSphereDatacenter.Name, vsphere.ClusterNameLabel: spec.Cluster.Name}); err != nil {
+			return controller.Result{}, errors.Wrap(err, "listing current Vsphere Deployment Zones")
+		}
+	
+		var allErrs []error
+	
+		for _, m := range existingVSphereDeploymentZones.Items {
+			if !desiredFailureDeploymentNames.Contains(m.Name) {
+				if err := r.client.Delete(ctx, &m); err != nil {
+					allErrs = append(allErrs, err)
+				}
+			}
+		}
+	
```

Comment:
Should we check for deletion of these resources or not needed as CAPV handles it?